Source/HTTP/WinHttp/winhttp_http_task.cpp

```diff
@@ -785,11 +962,79 @@ http_internal_wstring flatten_http_headers(_In_ const http_header_map& headers)
     return flattened_headers;
 }
 
+HRESULT winhttp_http_task::query_security_information(
+    http_internal_wstring wUrlHost
+    )
+{
+#if HC_PLATFORM == HC_PLATFORM_GDK
+
+    XAsyncBlock asyncBlock{};
+    asyncBlock.queue = m_env->GetImmediateQueue();
+    HRESULT hr = XNetworkingQuerySecurityInformationForUrlUtf16Async(wUrlHost.c_str(), &asyncBlock);
+    if (FAILED(hr))
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "winhttp_http_task [ID %llu] [TID %ul] XNetworkingQuerySecurityInformationForUrlUtf16Async 0x%0.8x", TO_ULL(HCHttpCallGetId(m_call)), GetCurrentThreadId(), hr);
+        return hr;
+    }
+
+    hr = XAsyncGetStatus(&asyncBlock, true);
+    if (FAILED(hr))
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "winhttp_http_task [ID %llu] [TID %ul] XNetworkingQuerySecurityInformationForUrlUtf16Async status 0x%0.8x", TO_ULL(HCHttpCallGetId(m_call)), GetCurrentThreadId(), hr);
+        return hr;
+    }
+
+    size_t securityInformationBufferByteCount{ 0 };
+    hr = XNetworkingQuerySecurityInformationForUrlUtf16AsyncResultSize(&asyncBlock, &securityInformationBufferByteCount);
+    if (FAILED(hr))
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "winhttp_http_task [ID %llu] [TID %ul] XNetworkingQuerySecurityInformationForUrlUtf16AsyncResultSize 0x%0.8x", TO_ULL(HCHttpCallGetId(m_call)), GetCurrentThreadId(), hr);
+        return hr;
+    }
+
+    assert(securityInformationBufferByteCount > 0);
+    m_securityInformation = nullptr;
+    m_securityInformationBuffer.resize(securityInformationBufferByteCount);
+
+    size_t bytesUsed;
+    hr = XNetworkingQuerySecurityInformationForUrlUtf16AsyncResult(
+        &asyncBlock,
+        m_securityInformationBuffer.size(),
+        &bytesUsed,
+        m_securityInformationBuffer.data(),
+        &m_securityInformation);
+    if (FAILED(hr))
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "winhttp_http_task [ID %llu] [TID %ul] XNetworkingQuerySecurityInformationForUrlUtf16AsyncResult 0x%0.8x", TO_ULL(HCHttpCallGetId(m_call)), GetCurrentThreadId(), hr);
+        return hr;
+    }
+#endif
+
+    return S_OK;
+}
+
 HRESULT winhttp_http_task::connect(
     _In_ const xbox::httpclient::Uri& cUri
     )
 {
-    if (!m_hSession)
+    HRESULT hr = S_OK;
+    m_isSecure = cUri.IsSecure();
+    unsigned int port = cUri.IsPortDefault() ?
+        (m_isSecure ? INTERNET_DEFAULT_HTTPS_PORT : INTERNET_DEFAULT_HTTP_PORT) :
+        cUri.Port();
+    http_internal_wstring wUrlHost = utf16_from_utf8(cUri.Host());
+
+#if HC_PLATFORM != HC_PLATFORM_GDK
+    uint32_t enabledHttpSecurityProtocolFlags = HC_PERFORM_ENV::GetDefaultHttpSecurityProtocolFlagsForWin7();
+#else
+    hr = query_security_information(wUrlHost);
+    if (FAILED(hr))
+        return hr;
+    uint32_t enabledHttpSecurityProtocolFlags = m_securityInformation->enabledHttpSecurityProtocolFlags;
```

Comment:
> m_securityInformation->enabledHttpSecurityProtocolFlags [](start = 47, length = 56)

 Minor thing, but reading the code it wasn't immediately clear how this get set. If this is the only field of m_securityInformation you need, maybe just return it as an _Out_ param from query_security_information rather than storing it as a member of the win_http_task?

Reply:
Its actually used in the callback as well in:

#if HC_PLATFORM == HC_PLATFORM_GDK
            case WINHTTP_CALLBACK_STATUS_SENDING_REQUEST:
            {
                callback_status_sending_request(hRequestHandle, pRequestContext, statusInfo);
                break;
            }
#endif // HC_PLATFORM_GDK